### ISSUE TITLE
Support for .mjs files and native ESM

### ISSUE THREAD

**Issue by pi0** [nuxt/nuxt.js#1935]
Most of the code in nuxt **projects** can now be written in modern ES6 syntax. Even more is going to be supported when 8 is next LTS (Such as async/await). However, there are parts currently are not fully ES6 compatible because are being required directly with node and thus depends on pure node functionalities. Also using babel-register to polyfill this gap may introduce many many inconsistencies and perf/stability issues which is not ideal. Current parts which can potentially leverage `.mjs`:

- Modules (to be used without any transpiler like roll-up)
- Server Middleware
- `nuxt.config.js`
- [potentailly] Nuxt.js core source code itself for perf and maintenance reasons

This [article](http://2ality.com/2017/09/native-esm-node.html) is useful for getting started with native esm. Currently, esm support can be enabled behind a flag so we would delay this feature for Nuxt until it gets stable enough.

PS: Any ideas, feedbacks, and contribution are welcome. our only limitation is that PRs should target v2 as this is almost breaking change in the entire ecosystem.

XREFs:
- #1912
- #2087

**Comment by pi0** [comment]
I think we can experiment most of the parts by using something like this to start Node.js process:

```bash
node --experimental-modules node_modules/nuxt --config nuxt.config.esm
```

Again [this](http://2ality.com/2017/09/native-esm-node.html) article would be really helpful to get started :)

**Comment by thgh** [comment]
I'm experimenting with `@std/esm` to share code between client & server. How about wrapping all serverMiddleware code with this?

```js
// serverMiddleware.js
const { Router } = require('express')
const router = Router()

const esm = require('@std/esm')(module, { esm: 'js', cjs: true })

router.use(esm('./some-api-module').router)
router.use(esm('./another-one').router)

module.exports = router
```

```js
// some-api-module.js
import { Router } from 'express'
export const router = Router()

router.post('/...')
```

**Comment by pi0** [comment]
Shipped! (#1935)
